Improve semantic indentation rules to be more consistent with cljfmt

rrudakov opened the pull request:
Example:

<img width="561" alt="image" src="https://github.com/user-attachments/assets/27665e0f-7dcc-41a5-9521-92e3001cafdc" />

Before submitting a PR mark the checkboxes for the items you've done (if you
think a checkbox does not apply, then leave it unchecked):

- [x] The commits are consistent with our [contribution guidelines][1].
- [x] You've added tests (if possible) to cover your change(s). Bugfix, indentation, and font-lock tests are extremely important!
- [x] You've run `M-x checkdoc` and fixed any warnings in the code you've written.
- [x] You've updated the changelog (if adding/changing user-visible functionality).
- [ ] You've updated the readme (if adding/changing user-visible functionality).

Thanks!

[1]: https://github.com/clojure-emacs/clojure-ts-mode/blob/master/CONTRIBUTING.md

bbatsov commented:
The proposal seems reasonable to me - from what I gather the essence here is that we want to treat differently forms depending on the position of their `body/forms` param. 

Might be a good idea to mention this somewhere in the README, and I'm also wondering if we should make this configurable like in `clojure-mode` (there you can add extra forms to the various indentation categories, so you'd get consistent indentation for your macros as well). 

Making this configurable is also need for CIDER to be able to insert indentation rules itself, based on forms indentation specs.

bbatsov commented on clojure-ts-mode.el:
```diff
@@ -746,33 +746,61 @@ The possible values for this variable are
+(defvar clojure-ts--symbols-block-1-regexp
+  (eval-and-compile
+    (rx (and line-start
+             (or "defprotocol"
```
Basically, I'm suggesting to make lists like this one separate vars that users can modify, as they will likely have macros in their own code that need to be indented in a non-standard way.

rrudakov replied:
Done.

bbatsov commented on clojure-ts-mode.el:
```diff
@@ -746,33 +746,61 @@ The possible values for this variable are
   "A regex to match symbols that are functions/macros with a body argument.
-Taken from cljfmt:
+Taken from cljfmt (all symbols with [[:block 0]] rule):
+https://github.com/weavejester/cljfmt/blob/fb26b22f569724b05c93eb2502592dfc2de898c3/cljfmt/resources/cljfmt/indents/clojure.clj")
```
We might go with some version tag here instead, so the URL is a bit more informative.

rrudakov replied:
Done.

rrudakov commented:
Thanks you for the review @bbatsov. I've refactored the code. It's less repetitive now and rules are customizable via `clojure-ts-semantic-indent-rules` variable. I also updated README. Please let me know what do you think.

bbatsov commented on clojure-ts-mode.el:
```diff
@@ -746,34 +761,129 @@ The possible values for this variable are
+    (when-let* ((node-after-bol (treesit-node-first-child-for-pos parent bol)))
+      (> (treesit-node-index node-after-bol) (1+ block)))))
+
+(defun clojure-ts--match-form-body (node parent bol)
```
I think you forgot to add a docstring here. (`checkdoc` complains for docstrings for all functions, not just the public ones)

rrudakov replied:
Forgot to add :) Now fixed, thanks!

bbatsov commented on README.md:
````diff
@@ -170,6 +170,20 @@ Set the var `clojure-ts-indent-style` to change it.
+custom rules as:
+
+```emacs-lisp
+(setopt clojure-ts-semantic-indent-rules '(("are" . (:block 1))))
````
Might be better to have several forms here, so it's clearer what the format is.

bbatsov replied:
I think it'd be good to explain also the various `block` and `inner` values here, similar to how it's done in `clojure-mode`'s docs.

rrudakov replied:
I've adapted some examples from `clojure-mode` documentation and added some explanation about the rules.

bbatsov commented on CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
   - Highlight named lambda functions properly.
   - Fix syntax highlighting for functions and vars with metadata on the previous
     line.
+- Improve semantic indentation rules to be more consistent with cljfmt.
```
I'd also mention the new defcustom in the changelog.

rrudakov replied:
Done.

bbatsov reviewed (approved):
Great work!
